### PR TITLE
MM - 43529 start freemium trial

### PR DESCRIPTION
**Pull request by pvev**
#### Summary
This PR adds the  ability to start a trial from cloud free workspaces.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-43529

#### Related Pull Requests
MM Server: https://github.com/mattermost/mattermost-server/pull/20163
Enterprise: https://github.com/mattermost/enterprise/pull/1197

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by pvev** [comment]
since PR https://github.com/mattermost/mattermost-webapp/pull/10283 was failing a lot to create a new spinwick server (maybe because I messed a lot with the stripe data) I created this new one.

File: components/cloud_start_trial/cloud_start_trial_btn.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    onClick,
+}: CloudStartTrialBtnProps) => {
+    const {formatMessage} = useIntl();
+    const dispatch = useDispatch<DispatchFunc>();
+
+    const [status, setLoadStatus] = useState(TrialLoadStatus.NotStarted);
+
+    const requestLicense = async (): Promise<TrialLoadStatus> => {
+        const productUpdated = await requestCloudTrial('start_trial_btn')();
+        if (!productUpdated) {
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/cloud_start_trial/cloud_start_trial_btn.tsx:52]: agree, but we are not doing anything with the catched error. We only need to know if the request was successful or not, and depending on that we change the button text. So what I did was modify the response so just return a boolean. Thanks for this catch, was a good one!

File: components/admin_console/feature_discovery/feature_discovery.tsx [hunk 2]
```diff
@@ -112,29 +116,40 @@ export default class FeatureDiscovery extends React.PureComponent<Props, State>
...
+        if (this.props.isCloud && !this.props.isCloudFreeEnabled) {
             primaryMessage = (
                 <FormattedMessage
                     id='admin.ldap_feature_discovery_cloud.call_to_action.primary'
                     defaultMessage='Subscribe now'
                 />
             );
         }
+        const {isCloud, isFreeTrial, isCloudFreeEnabled, hadPrevFreeTrial} = this.props;
+        const canRequestCloudFreTrial = isCloud && isCloudFreeEnabled && !isFreeTrial && !hadPrevFreeTrial;
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/admin_console/feature_discovery/feature_discovery.tsx:128]: ok!

File: components/learn_more_trial_modal/learn_more_trial_modal_step.tsx [hunk 3]
```diff
@@ -30,8 +43,39 @@ const LearnMoreTrialModalStep = (
...
+    let startTrialBtnMsg = formatMessage({id: 'start_trial.modal_btn.start_free_trial', defaultMessage: 'Start free 30-day trial'});
+
+    // Cloud conditions
+    const license = useSelector(getLicense);
+    const isCloud = license?.Cloud === 'true';
+    const isCloudFreeEnabled = useSelector(cloudFreeEnabled);
+
+    // close this modal once start trial btn is clicked and trial has started successfully
+    const dismissAction = async () => {
+        await dispatch(closeModal(ModalIdentifiers.LEARN_MORE_TRIAL_MODAL));
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/learn_more_trial_modal/learn_more_trial_modal_step.tsx:57]: ok! agree!

File: components/learn_more_trial_modal/learn_more_trial_modal_step.tsx [hunk 4]
```diff
@@ -30,8 +43,39 @@ const LearnMoreTrialModalStep = (
...
+    };
+
+    if (isCloud) {
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/learn_more_trial_modal/learn_more_trial_modal_step.tsx:60]: changed!

File: components/widgets/menu/menu_items/menu_cloud_trial.tsx [hunk 5]
```diff
@@ -2,53 +2,126 @@
...
     }
 
-    return (
-        <li
-            className={'MenuCloudTrial'}
-            role='menuitem'
-            id={id}
-        >
+    // menu content displayed when the workspace is running a trial, it depends also if cloudfree is enabled
+    const freeTrialContent = isCloudFreeEnabled ? (
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/widgets/menu/menu_items/menu_cloud_trial.tsx:58]: changed the comment so it makes more sense!

File: components/cloud_start_trial/cloud_start_trial_btn.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (pvev) [reply components/cloud_start_trial/cloud_start_trial_btn.tsx:52]: also with this catch I improved the unit tests to reflect both scenarios for the button, on successful trial request and on failed trial request

**Comment by neallred** [comment]
Testing locally, I had the wrong version of CWS so the start trial endpoint failed so I saw the fail state. Do we want to show the network response status in the button, or should we show fail state somewhere else (possibly red text below the button)?
text-in-button

**Comment by pvev** [comment]
> Testing locally, I had goofy data so the start trial endpoint failed so I saw the fail state. Do we want to show the network response status in the button, or should we show fail state somewhere else (possibly red text below the button)? text-in-button

The requestTrial(self hosted) shows the state in the button, I think is a good idea for keeping the UX consistent.

File: components/widgets/menu/menu_items/menu_cloud_trial.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (pvev) [reply components/widgets/menu/menu_items/menu_cloud_trial.tsx:59]: @neallred not sure if you noticed I refactored the comment and part of the code for this suggestion?

[the remark this answers is not in the document]
Comment (neallred) [reply components/widgets/menu/menu_items/menu_cloud_trial.tsx:59]: ahh, ok. This one looks good to me!

**Comment by neallred** [comment]
> The requestTrial(self hosted) shows the state in the button, I think is a good idea for keeping the UX consistent.

That makes sense to me.

File: components/widgets/menu/menu_items/menu_cloud_trial.tsx [hunk 6]
```diff
@@ -2,53 +2,127 @@
...
-        >
+    // this is the menu content displayed when the workspace is running a trial. It would display a different option depending if
+    // cloudFree is enabled or not.
+    const freeTrialContent = isCloudFreeEnabled ? (
+
+        // Here go the limits calculation and refactor of the menu option
+        <>
+            <FormattedMessage
+                id='menu.cloudFree.tryEnterprise'
+                defaultMessage='Are you making the most of your Enterprise trial? '
```
Comment (neallred) [inline components/widgets/menu/menu_items/menu_cloud_trial.tsx:65]: Looks like we need to create the id for this message. Right now, it is accidentally displaying the `Interested in a limitless plan with high-security features?` message.

needs-id

Comment (pvev) [reply components/widgets/menu/menu_items/menu_cloud_trial.tsx:65]: got it, nice catch!

**Review by marianunez — CHANGES_REQUESTED** [review]
Thanks @pvev! Overall looks good, some comments below but primarily we need to consider the case where they are already subscribed so we don't want to show any options to Start Trial

File: components/widgets/menu/menu_items/menu_cloud_trial.tsx [hunk 7]
```diff
@@ -2,53 +2,127 @@
...
-        <li
-            className={'MenuCloudTrial'}
-            role='menuitem'
-            id={id}
-        >
+    // this is the menu content displayed when the workspace is running a trial. It would display a different option depending if
+    // cloudFree is enabled or not.
+    const freeTrialContent = isCloudFreeEnabled ? (
+
+        // Here go the limits calculation and refactor of the menu option
```
[the remark this answers is not in the document]
Comment (neallred) [reply components/widgets/menu/menu_items/menu_cloud_trial.tsx:61]: Yeah, agree with that. In my limits modal branch, I started adding hooks for getting usage and limits: https://github.com/mattermost/mattermost-webapp/tree/MM-43596/components/common/hooks . When implemented, any sort of calculation could be placed there.

**Comment by marianunez** [comment]
@p

> Mattermost test server with CWS created! 🎉
> 
> Access here: https://mattermost-webapp-pr-10297.test.mattermost.cloud
> 
> Account Type	Username	Password
> Admin	[user-mattermost-webapp-pr-10297@example.mattermost.com](mailto:user-mattermost-webapp-pr-10297@example.mattermost.com)	Cws@10102020

@pvev heads up that the spinwick is back up but the trial promtps don't see to do anything right now

File: components/admin_console/feature_discovery/index.tsx [hunk 8]
```diff
@@ -19,10 +21,14 @@ import {GlobalState} from 'types/store';
...
 
 function mapStateToProps(state: GlobalState) {
+    const subscription = state.entities.cloud.subscription;
     return {
         stats: state.entities.admin.analytics,
         prevTrialLicense: state.entities.admin.prevTrialLicense,
         isCloud: isCloudLicense(getLicense(state)),
+        isCloudFreeEnabled: cloudFreeEnabled(state),
+        isFreeTrial: subscription?.is_free_trial === 'true',
+        hadPrevFreeTrial: subscription?.is_free_trial === 'false' && subscription?.trial_end_at > 0,
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/admin_console/feature_discovery/index.tsx:31]: done!

File: components/admin_console/feature_discovery/index.tsx [hunk 9]
```diff
@@ -19,10 +21,14 @@ import {GlobalState} from 'types/store';
 import FeatureDiscovery from './feature_discovery';
 
 function mapStateToProps(state: GlobalState) {
+    const subscription = state.entities.cloud.subscription;
     return {
         stats: state.entities.admin.analytics,
         prevTrialLicense: state.entities.admin.prevTrialLicense,
         isCloud: isCloudLicense(getLicense(state)),
+        isCloudFreeEnabled: cloudFreeEnabled(state),
+        isFreeTrial: subscription?.is_free_trial === 'true',
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/admin_console/feature_discovery/index.tsx:30]: ok, done!

File: components/admin_console/feature_discovery/feature_discovery.tsx [hunk 10]
```diff
@@ -112,29 +116,40 @@ export default class FeatureDiscovery extends React.PureComponent<Props, State>
...
                 />
             );
         }
+        const {isCloud, isFreeTrial, isCloudFreeEnabled, hadPrevFreeTrial} = this.props;
+        const canRequestCloudFreeTrial = isCloud && isCloudFreeEnabled && !isFreeTrial && !hadPrevFreeTrial;
+
+        const ctaTrialButton = canRequestCloudFreeTrial ? (
+            <CloudStartTrialBtn
+                message={Utils.localizeMessage('menu.cloudFree.tryEnterpriseFor30Days', 'Try Enterprise free for 30 days')}
+                telemetryId={'start_cloud_trial_after_completing_steps'}
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/admin_console/feature_discovery/feature_discovery.tsx:133]: yes! nice catch.

File: components/cloud_start_trial/cloud_start_trial_btn.tsx [hunk 11]
```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {useState} from 'react';
+
+import {useIntl} from 'react-intl';
+
+import {useDispatch} from 'react-redux';
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/cloud_start_trial/cloud_start_trial_btn.tsx:8]: ok, done!

File: components/cloud_start_trial/cloud_start_trial_btn.tsx [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+import {DispatchFunc} from 'mattermost-redux/types/actions';
+
+import {requestCloudTrial} from 'actions/cloud';
+
+import {trackEvent} from 'actions/telemetry_actions';
+
+import {openModal} from 'actions/views/modals';
+
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/cloud_start_trial/cloud_start_trial_btn.tsx:17]: ok, done!

File: components/cloud_start_trial/cloud_start_trial_btn.tsx [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+enum TrialLoadStatus {
+    NotStarted = 'NOT_STARTED',
+    Started = 'STARTED',
+    Success = 'SUCCESS',
+    Failed = 'FAILED',
+    Embargoed = 'EMBARGOED',
+}
+
+const CloudStartTrialBtn = ({
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/cloud_start_trial/cloud_start_trial_btn.tsx:39]: ok, done!

File: components/cloud_start_trial/cloud_start_trial_btn.tsx [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    telemetryId,
+    linkStyle,
+    onClick,
+}: CloudStartTrialBtnProps) => {
+    const {formatMessage} = useIntl();
+    const dispatch = useDispatch<DispatchFunc>();
+
+    const [status, setLoadStatus] = useState(TrialLoadStatus.NotStarted);
+
+    const requestLicense = async (): Promise<TrialLoadStatus> => {
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/cloud_start_trial/cloud_start_trial_btn.tsx:50]: agree, done!

File: components/learn_more_trial_modal/learn_more_trial_modal_step.test.tsx [hunk 15]
```diff
@@ -17,20 +23,51 @@ describe('components/learn_more_trial_modal/learn_more_trial_modal_step', () =>
...
+    const state = {
+        entities: {
+            admin: {
+                prevTrialLicense: {
+                    IsLicensed: 'false',
+                },
+            },
+            general: {
+                license: {
+                    IsLicensed: 'false',
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/learn_more_trial_modal/learn_more_trial_modal_step.test.tsx:35]: added three more tests to cover the cloud scenarios.

File: components/admin_console/feature_discovery/feature_discovery.tsx [hunk 16]
```diff
@@ -112,29 +116,40 @@ export default class FeatureDiscovery extends React.PureComponent<Props, State>
...
+        if (this.props.isCloud && !this.props.isCloudFreeEnabled) {
             primaryMessage = (
                 <FormattedMessage
                     id='admin.ldap_feature_discovery_cloud.call_to_action.primary'
                     defaultMessage='Subscribe now'
                 />
             );
         }
+        const {isCloud, isFreeTrial, isCloudFreeEnabled, hadPrevFreeTrial} = this.props;
+        const canRequestCloudFreeTrial = isCloud && isCloudFreeEnabled && !isFreeTrial && !hadPrevFreeTrial;
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/admin_console/feature_discovery/feature_discovery.tsx:128]: is my understanding that feature discovery will only show when is not paid customer (cloud) or has not license (self hosted). However it makes sense to add that validation for the Cloud Trial Menu.

File: components/onboarding_tasklist/onboarding_tasklist_completed.tsx [hunk 17]
```diff
@@ -119,8 +121,19 @@ const Completed = (props: Props): JSX.Element => {
...
+    // Cloud conditions
+    const subscription = useSelector((state: GlobalState) => state.entities.cloud.subscription);
+    const isCloud = license?.Cloud === 'true';
+    const isFreeTrial = subscription?.is_free_trial === 'true';
+    const hadPrevFreeTrial = subscription?.is_free_trial === 'false' && subscription?.trial_end_at > 0;
+    const isCloudFreeEnabled = useSelector(cloudFreeEnabled);
+
     // Show this CTA if the instance is currently not licensed and has never had a trial license loaded before
-    const showStartTrialBtn = (isCurrentLicensed === 'false' && isPrevLicensed === 'false');
+    // if Cloud, show if isCloudFreeEnabled and is not in trial and had never been on trial
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/onboarding_tasklist/onboarding_tasklist_completed.tsx:132]: done!

File: components/widgets/menu/menu_items/menu_cloud_trial.tsx [hunk 18]
```diff
@@ -2,53 +2,127 @@
...
+import {TrialPeriodDays, ModalIdentifiers} from 'utils/constants';
 
 import './menu_item.scss';
+import CloudStartTrialBtn from 'components/cloud_start_trial/cloud_start_trial_btn';
 
 type Props = {
     id: string;
 }
-const MenuCloudTrial: React.FC<Props> = ({id}: Props) => {
+const MenuCloudTrial = ({id}: Props) => {
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/widgets/menu/menu_items/menu_cloud_trial.tsx:27]: added several new test to cover all scenarios

File: components/admin_console/feature_discovery/feature_discovery.tsx [hunk 19]
```diff
@@ -112,29 +117,46 @@ export default class FeatureDiscovery extends React.PureComponent<Props, State>
...
+            isCloudFreeEnabled,
+            hadPrevCloudTrial,
+            isCloudFreePaidSubscription,
+        } = this.props;
+        const canRequestCloudFreeTrial = isCloud && isCloudFreeEnabled && !isCloudTrial && !hadPrevCloudTrial && !isCloudFreePaidSubscription;
+
+        const ctaTrialButton = canRequestCloudFreeTrial ? (
+            <CloudStartTrialButton
+                message={Utils.localizeMessage('menu.cloudFree.tryEnterpriseFor30Days', 'Try Enterprise free for 30 days')}
+                telemetryId={'start_cloud_trial_feature_discovery'}
```
Comment (marianunez) [inline components/admin_console/feature_discovery/feature_discovery.tsx:140]: Is there any way to know what feature this is for? Maybe the title? Would be useful to have that in the telemetry

Comment (pvev) [reply components/admin_console/feature_discovery/feature_discovery.tsx:140]: makes sense, let me check!

**Comment by dontoisme** [comment]
@marianunez confirming that the fail state for trial requests is expected at this time yes?
Screen Shot 2022-05-13 at 9 35 56 AM

**Comment by pvev** [comment]
> @marianunez confirming that the fail state for trial requests is expected at this time yes? Screen Shot 2022-05-13 at 9 35 56 AM

@dontoisme @marianunez it seems to be that the data in stripe is not matching exactly what we need to request the trial successfully for this installation. At first glance it seems that it is creating the product to be Enterprise instead of the Cloud Free.

**Comment by marianunez** [comment]
> @dontoisme @marianunez it seems to be that the data in stripe is not matching exactly what we need to request the trial successfully for this installation. At first glance it seems that it is creating the product to be Enterprise instead of the Cloud Free.

Yes, it seems the spinwick is starting in Enteprise (not even in Trial) and not Freemium even though the feature flag is enabled properly. @nickmisasi can we double check that the recent changes done to the internal endpoint is not affecting this

**Comment by nickmisasi** [comment]
@marianunez this isn't a regression from the recent changes - the internal route just still defaults to Enterprise. Fix here: https://github.com/mattermost/customer-web-server/pull/776

In the meantime, I've manually shifted the subscription for this spinwick into Freemium.

**Comment by pvev** [comment]
> @marianunez this isn't a regression from the recent changes - the internal route just still defaults to Enterprise. Fix here: [mattermost/customer-web-server#776](https://github.com/mattermost/customer-web-server/pull/776)
> 
> In the meantime, I've manually shifted the subscription for this spinwick into Freemium.

thank you @nickmisasi , it seems that it worked.
Captura de pantalla 2022-05-13 a las 19 51 09

@marianunez One thing to note is that I noticed the trial was set to Enterprise but to last 14 days.  (just noticed stripe displays to me in spanish :D) https://dashboard.stripe.com/test/subscriptions/sub_1Kz0lYI67GP2qpb4s7G7IBWn/edit

Captura de pantalla 2022-05-13 a las 19 50 50

**Comment by nickmisasi** [comment]
The change subscription handler that drives this looks at the feature flag value for `TrialPeriodInDays` - and falls back on the 14 day trial. `TrialPeriodInDays` is not set in split.io right now, so it's using the 14 day period.

**Comment by pvev** [comment]
> The change subscription handler that drives this looks at the feature flag value for `TrialPeriodInDays` - and falls back on the 14 day trial. `TrialPeriodInDays` is not set in split.io right now, so it's using the 14 day period.

ok, got it. Thanks @nickmisasi .

File: components/admin_console/feature_discovery/feature_discovery.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (marianunez) [reply components/admin_console/feature_discovery/feature_discovery.tsx]: > is my understanding that feature discovery will only show when is not paid customer (cloud) or has not license (self hosted). 

Yes, good point!

File: components/onboarding_tasklist/onboarding_tasklist_completed.tsx [hunk 20]
```diff
@@ -119,8 +122,20 @@ const Completed = (props: Props): JSX.Element => {
     const isPrevLicensed = prevTrialLicense?.IsLicensed;
     const isCurrentLicensed = license?.IsLicensed;
 
+    // Cloud conditions
+    const subscription = useSelector((state: GlobalState) => state.entities.cloud.subscription);
+    const isCloud = license?.Cloud === 'true';
+    const isFreeTrial = subscription?.is_free_trial === 'true';
+    const hadPrevCloudTrial = subscription?.is_free_trial === 'false' && subscription?.trial_end_at > 0;
+    const isCloudFreeEnabled = useSelector(cloudFreeEnabled);
+    const isCloudFreePaidSubscription = isCloud && isCloudFreeEnabled && license?.SkuShortName !== LicenseSkus.Starter && !isFreeTrial;
```
Comment (marianunez) [inline components/onboarding_tasklist/onboarding_tasklist_completed.tsx:131]: nit: it's confusing Cloud Free and Paid in the name - is it free or paid? 🤣 

```suggestion
    const isCloudPaidSubscription = isCloud && isCloudFreeEnabled && license?.SkuShortName !== LicenseSkus.Starter && !isFreeTrial;
```

**Review by marianunez — APPROVED** [review]
Great work @pvev 💯

**Comment by marianunez** [comment]
@pvev Doing a pass on the test server here, it seems that it may not be picking up your changes in the server/enterprise branch because the endpoint is doing a 404. If those are ready, I would suggest to merge them so it can be picked up by the spinwick and avoid the flakyness of the matterwick. 

Additionally, I did notice a few UX details:

1. The Start Trial in the Feature Discovery page seems to be missing style:
Screen Shot 2022-05-17 at 7 06 28 PM

2. Minor misaligment in the Product Switcher menu for the link to Start Trial line: 
 
Screen Shot 2022-05-17 at 7 09 52 PM

**Comment by dontoisme** [comment]
Screen Shot 2022-05-18 at 2 27 07 PM
When starting a trial, I'm assuming that the 14 day aspect of it is configurable and will be updated to 30 days at a later time?

**Comment by pvev** [comment]
> I'm assuming that the 14 day (at the top of the screenshot) aspect of it is configurable and will be updated to 30 days at a later time?

Yes, it will change for the Freemium feature. @marianunez we need to keep in mind for this that the code in some parts makes reference to 14 days left, 3 days left and 1 day aka last day. How are we going to handle this for the 30 days period?

**Review by marianunez — APPROVED** [review]

**Comment by marianunez** [comment]
Created issues for the open of the modal and the 14 day harcoded references [MM-44420](https://mattermost.atlassian.net/browse/MM-44420) [MM-44419](https://mattermost.atlassian.net/browse/MM-44419)

**Comment by pvev** [comment]
> Successfully triggered e2e testing! https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/196320

e2e results :) https://automation-dashboard.vercel.app/cycles/0e8ea7d7-b2c5-4987-8890-fc605a48bfec

**Review by stevemudie — APPROVED** [review]
Looks good @pvev!

I was able to start the trial as expected.

**Comment by pvev** [comment]
> Looks good @pvev!
> 
> I was able to start the trial as expected.

awesome! thanks @steve!

**Comment by pvev** [comment]
@marianunez making the link to open the modal was a really small change which I added fixing the merge conflicts. So we can keep MM-44419 to track the changes to show the View Plans new modal.


https://user-images.githubusercontent.com/10082627/169400570-a661c3dd-29aa-4a32-ab41-e526d9851790.mp4

**Comment by marianunez** [comment]
> @marianunez making the link to open the modal was a really small change which I added fixing the merge conflicts. So we can keep MM-44419 to track the changes to show the View Plans new modal.

Awesome, sounds good to me 👍

**Comment by marianunez** [comment]
Can we cherry-pick to `cloud`? This will not be available in the next cloud release because it's behind a feature flag but would like to have in the cloud test servers for our integration testing cc/ @amyblais

**Pull request merged**